QUICK-FIX Add tests for filling assessment GCA and LCA

abotalov opened the pull request:
# Issue description

*Explain the issue that the changes in the pull request solve.*

# Steps to test the changes

*Include steps to test or reproduce (in case of a bugfix) the changes in the pull request. This section should help the reviewer verify your changes.*

# Solution description

*Briefly explain the solution or add a link to the technical design document (this will help the reviewer confirm the code does what you intend). Feel free to also mention alternative approaches that didn't work out.*

# Sanity checklist

- [ ] I have clicked through the app to make sure my changes work and not break the app.
- [ ] I have applied the correct milestone and labels.
- [ ] My changes fix the issue described in the description (and do nothing else). 🤞
- [ ] My changes are covered by tests.
- [ ] My changes follow our [performance guidelines](https://github.com/google/ggrc-core/blob/dev/docs/source/contributing/performance.rst).
- [ ] My changes follow our [js](https://github.com/google/ggrc-core/blob/dev/docs/source/contributing/javascript.rst) and/or [python](https://github.com/google/ggrc-core/blob/dev/docs/source/contributing/python.rst) guidelines.
- [ ] My commits follow our [commit guidelines](https://github.com/google/ggrc-core/blob/dev/docs/source/contributing/git/how_to_write_a_commit_message.rst).

<!-- If your PR includes a migration include the additional checklist items
# Migration checklist
- [ ] Migration passes all checks from our [PR review guidelines](https://github.com/google/ggrc-core/blob/dev/docs/source/contributing/git/reviewing_pull_requests.rst#reviewing-a-pr-containing-database-migration-scripts)
-->

# PR Review checklist

- [ ] The changes fix the issue and don't cause any apparent regressions.
- [ ] Labels and milestone are correctly set.
- [ ] The solution description matches the changes in the code.
- [ ] There is no apparent way to improve the performance & design of the new code.
- [ ] The pull request is opened against the correct base branch.
- [ ] Upon merging, the Jira ticket's fixversion is correctly set and the ticket is moved to "QA - In Progress".

<!-- If your code is not finished yet can include a TODO check list
# TODO

- [ ] First item on the TODO
-->

SergeyRabtsau commented on test/selenium/src/lib/entities/entities_factory.py:
```diff
@@ -44,7 +45,7 @@ def _create_random_obj(self, is_add_rest_attrs):
     (method for overriding).
     """
-    raise NotImplementedError
+    pass
```
**Question** @abotalov why do change it? As far I understand this is abstract method, in case if you want to change to it I suggest to use `abc` module

abotalov replied:
This method doesn't need to be overriden if `_set_attrs` is overriden.

SergeyRabtsau commented on test/selenium/src/lib/entities/entities_factory.py:
```diff
@@ -504,6 +490,41 @@ def _create_random_obj(self, is_add_rest_attrs):
 
 
+class AsmtTemplateWithDropdownFactory(AssessmentTemplatesFactory):
+  """Class for assesment template with Dropdown option"""
```
**DocString** I suggest to use common approach and put . (dot) at the end of doc string.

abotalov replied:
Fixed

SergeyRabtsau commented on test/selenium/src/lib/service/rest_facade.py:
```diff
@@ -37,74 +34,38 @@ def create_audit(program, **attrs):
 
 
-def create_assessment_from_template(audit, template, **attrs):
+def create_asmt_from_template(audit, asmt_template, control):
```
We can create asmt not only based on control, so please make sure that we handle this in correct way.

abotalov replied:
Fixed

abotalov commented:
Added some changes:
* Addressed review comments
* Changed `test_fill_asmt_gcas_in_popup` to run under several roles
* Added `test_fill_asmt_gcas_inline` to fill GCAs inline
* Some changes in method names and docstrings

abotalov commented on test/selenium/src/lib/page/widget/page_elements.py:
```diff
@@ -81,3 +92,322 @@ class CommentArea(object):
+    """Gets value of inline LCA field"""
+    return self._input.text
+
+  def set_lcas_from_inline(self, value):
```
Goal of code reuse is to adhere to DRY principle:
> Every piece of knowledge must have a single, unambiguous, authoritative representation within a system

In my opinion code inside TextCAActionsStrategy and RichTextCAActionsStrategy is mostly the same by accident. UI for text and rich text may be changed independently so changes to TextCAActionsStrategy and RichTextCAActionsStrategy would also have to be independent.
Also making RichTextCAActionsStrategy a subclass of TextCAActionsStrategy may be violation of LSP.
See also https://en.wikipedia.org/wiki/Rule_of_three_(computer_programming)

abotalov commented on test/selenium/src/lib/page/widget/page_elements.py:
```diff
@@ -2,10 +2,37 @@
+    save_btn.wait_until_not(lambda e: e.exists)
+    # Wait for JS to work, there are no DOM changes and HTTP requests
+    # during some period (GGRC-5891).
+    time.sleep(1)
```
FE investigation is required to find out. I decided to create a ticket so there will be no waiting.
I made some checks, 500 ms is not enough. 1 second is enough for DOM changes to start.

SergeyRabtsau replied:
ok, clear

abotalov commented on test/selenium/src/lib/page/widget/page_elements.py:
```diff
@@ -81,3 +92,322 @@ class CommentArea(object):
+        self._ca_strategy.set_lcas_from_inline(value)
+      # Wait for JS to work, there are no DOM changes and HTTP requests
+      # during some period (GGRC-5891).
+      time.sleep(1)
```
Please see a comment above.

abotalov commented on test/selenium/src/lib/page/widget/page_elements.py:
```diff
@@ -81,3 +92,322 @@ class CommentArea(object):
+      js_el = self._root.element(class_name="inline-edit-control")
+    else:
+      js_el = self._root.element(class_name="form-field__content")
+    js_type = js_el.browser.execute_script(
```
I'm not sure this line would be useful in other places.

abotalov commented on test/selenium/src/lib/service/rest_facade.py:
```diff
@@ -37,74 +34,38 @@ def create_audit(program, **attrs):
+
+
+def create_gcad(**attrs):
+  """Create global CADs for all types"""
```
Fixed

abotalov commented on test/selenium/src/lib/page/widget/page_mixins.py:
```diff
@@ -26,6 +26,10 @@ def _comment_area(self):
     return page_elements.CommentArea(self._browser)
 
+  def _edit_popup(self):
+    """Return Edit popup page element"""
```
Fixed

abotalov commented on test/selenium/src/lib/page/widget/page_elements.py:
```diff
@@ -81,3 +92,322 @@ class CommentArea(object):
+    super(RichTextCAActionsStrategy, self).__init__(*args)
+    self._input = self._root.element(class_name="ql-editor")
+
+  def get_lcas_from_inline(self):
```
Fixed

abotalov commented on test/selenium/src/lib/page/widget/page_elements.py:
```diff
@@ -81,3 +92,322 @@ class CommentArea(object):
+    return self._input.value
+
+  def set_lcas_from_inline(self, value):
+    """Sets value of inline LCA field"""
```
Fixed

abotalov commented on test/selenium/src/lib/page/widget/page_elements.py:
```diff
@@ -2,10 +2,37 @@
+    self._root.element(class_name="fa-pencil").js_click()
+
+  def confirm(self):
+    """Save changes via inline edit"""
```
Fixed

abotalov commented on test/selenium/src/lib/page/widget/page_elements.py:
```diff
@@ -2,10 +2,37 @@
+    self._root = root
+
+  def open(self):
+    """Open inline edit"""
```
Fixed

abotalov commented on test/selenium/src/lib/page/widget/page_elements.py:
```diff
@@ -2,10 +2,37 @@
+
+
+class InlineEdit(object):
+  """Inline edit page element"""
```
Fixed

abotalov commented on test/selenium/src/lib/page/widget/info_widget.py:
```diff
@@ -584,25 +514,36 @@ def get_info_widget_obj_scope(self):
+    self.fill_ca_values(custom_attributes, is_global=False, is_inline=True)
+
+  def fill_global_cas_inline(self, custom_attributes):
+    """Fills GCAs inline"""
```
Fixed

abotalov commented on test/selenium/src/lib/page/widget/info_widget.py:
```diff
@@ -584,25 +514,36 @@ def get_info_widget_obj_scope(self):
+    return {k.upper(): v for k, v in custom_attributes.iteritems()}
+
+  def fill_local_cas(self, custom_attributes):
+    """Fills local CAs on asmt page"""
```
Fixed

abotalov commented on test/selenium/src/lib/page/widget/info_widget.py:
```diff
@@ -584,25 +514,36 @@ def get_info_widget_obj_scope(self):
+    return self.get_custom_attributes()
+
+  def local_custom_attributes(self):
+    """Switch to tab with LCA and return their values"""
```
Fixed

abotalov commented on test/selenium/src/lib/page/widget/info_widget.py:
```diff
@@ -584,25 +514,36 @@ def get_info_widget_obj_scope(self):
-      self.cas_scope_txt.update(self.lcas_scope_txt)
-      self._extend_list_all_scopes(self.cas_lbl_txt, self.cas_scope_txt)
+  def global_custom_attributes(self):
+    """Switch to tab with GCA and return their values"""
```
Fixed

abotalov commented on test/selenium/src/lib/page/widget/info_widget.py:
```diff
@@ -211,121 +211,62 @@ def get_header_and_value_txt_from_people_scopes(self, header_text):
+    self.fill_ca_values(custom_attributes, is_global=True, is_inline=False)
+
+  def fill_global_cas_inline(self, custom_attributes):
+    """Fills GCAs inline"""
```
Fixed

abotalov commented on test/selenium/src/lib/page/widget/info_widget.py:
```diff
@@ -211,121 +211,62 @@ def get_header_and_value_txt_from_people_scopes(self, header_text):
+    return custom_attributes
+
+  def fill_global_cas_in_popup(self, custom_attributes):
+    """Fill GCAs using Edit popup"""
```
Fixed

abotalov commented on test/selenium/src/lib/page/widget/info_widget.py:
```diff
@@ -211,121 +211,62 @@ def get_header_and_value_txt_from_people_scopes(self, header_text):
-    """Get text of all CAs headers and values elements scopes and convert it to
-    dictionary. If 'is_gcas_not_lcas' then get GCAs, if not 'is_gcas_not_lcas'
+  def global_custom_attributes(self):
+    """Return GCA values"""
```
Fixed

abotalov commented on test/selenium/src/lib/entities/entities_factory.py:
```diff
@@ -504,6 +490,41 @@ def _create_random_obj(self, is_add_rest_attrs):
+
+  @classmethod
+  def generate_cad(cls, **attrs):
+    """Create CAD for asmt template"""
```
Fixed

abotalov commented on test/selenium/src/lib/entities/entities_factory.py:
```diff
@@ -477,6 +453,16 @@ def clone(asmt_tmpl, count_to_clone=1):
 
+  @staticmethod
+  def generate_cad(**attrs):
+    """Generate CAD for asmt template"""
```
Fixed

abotalov commented on test/selenium/conftest.py:
```diff
@@ -594,6 +595,20 @@ def audits(program):
+
+@pytest.fixture()
+def gcads_for_asmt():
+  """Create GCADs of all types for Assessment"""
```
Fixed

abotalov commented on test/selenium/conftest.py:
```diff
@@ -594,6 +595,20 @@ def audits(program):
 
+@pytest.fixture()
+def assessment(audit):
+  """Create an assessment within audit"""
```
Fixed

abotalov commented on test/selenium/src/lib/page/widget/page_elements.py:
```diff
@@ -81,3 +92,322 @@ class CommentArea(object):
+    self._input = self._root.text_field(class_name="text-field")
+
+  def get_lcas_from_inline(self):
+    """Gets value of inline LCA field"""
```
Fixed

abotalov commented on test/selenium/src/lib/page/widget/page_elements.py:
```diff
@@ -81,3 +92,322 @@ class CommentArea(object):
+    self._fill_input_field(value)
+
+  def _fill_input_field(self, value):
+    """Fills input field"""
```
Changed to raise `NotImplementedError`

abotalov commented on test/selenium/src/lib/page/widget/page_elements.py:
```diff
@@ -81,3 +92,322 @@ class CommentArea(object):
+    self._inline_edit.confirm()
+
+  def set_gcas_from_popup(self, value):
+    """Sets value of GCA field in popup"""
```
Fixed

abotalov commented on test/selenium/src/lib/page/widget/page_elements.py:
```diff
@@ -81,3 +92,322 @@ class CommentArea(object):
+    return self._root.element(class_name="inline__content").text
+
+  def set_gcas_from_inline(self, value):
+    """Sets value of inline GCA field"""
```
Fixed

abotalov commented on test/selenium/src/lib/page/widget/page_elements.py:
```diff
@@ -81,3 +92,322 @@ class CommentArea(object):
+    self._inline_edit = InlineEdit(self._root)
+
+  def get_gcas_from_inline(self):
+    """Gets value of inline GCA field"""
```
Fixed

abotalov commented on test/selenium/src/lib/page/widget/page_elements.py:
```diff
@@ -81,3 +92,322 @@ class CommentArea(object):
+
+
+class CAActionsStrategy(object):
+  """Parent class for custom attribute actions"""
```
Fixed

abotalov commented on test/selenium/src/lib/page/widget/page_elements.py:
```diff
@@ -81,3 +92,322 @@ class CommentArea(object):
+
+  @property
+  def ca_type(self):
+    """Returns type of CA element"""
```
Fixed

abotalov commented on test/selenium/src/lib/page/widget/page_elements.py:
```diff
@@ -81,3 +92,322 @@ class CommentArea(object):
+    return selenium_utils.filter_by_text(self._ca_manager.all_labels, title)
+
+  def _get_root(self, label_el):
+    """Finds root element of CA by label element"""
```
Fixed

abotalov commented on test/selenium/src/lib/page/widget/page_elements.py:
```diff
@@ -81,3 +92,322 @@ class CommentArea(object):
+    }[self.ca_type](self._root, self._label_el)
+
+  def _get_label_el(self, title):
+    """Finds label element by title"""
```
Fixed

abotalov commented on test/selenium/src/lib/page/widget/page_elements.py:
```diff
@@ -81,3 +92,322 @@ class CommentArea(object):
+
+
+class CustomAttribute(object):
+  """Custom attribute page element"""
```
Fixed

abotalov commented on test/selenium/src/lib/page/widget/page_elements.py:
```diff
@@ -81,3 +92,322 @@ class CommentArea(object):
+    return [el.text for el in self.all_labels]
+
+  def find_ca_elem_by_title(self, title):
+    """Finds CA page element based on title"""
```
Fixed

abotalov commented on test/selenium/src/lib/page/widget/page_elements.py:
```diff
@@ -81,3 +92,322 @@ class CommentArea(object):
+    return list(elements)
+
+  def all_ca_titles(self):
+    """Returns all CA titles"""
```
Fixed

abotalov commented on test/selenium/src/lib/page/widget/page_elements.py:
```diff
@@ -81,3 +92,322 @@ class CommentArea(object):
+    self.all_labels = self._get_all_labels()
+
+  def _get_all_labels(self):
+    """Returns list of all labels"""
```
Fixed

abotalov commented on test/selenium/src/lib/page/widget/page_elements.py:
```diff
@@ -81,3 +92,322 @@ class CommentArea(object):
+    self.modal = container.element(class_name="modal-wide")
+
+  def close_and_save(self):
+    """Save and close edit popup"""
```
Fixed

abotalov commented on test/selenium/src/lib/page/widget/page_elements.py:
```diff
@@ -81,3 +92,322 @@ class CommentArea(object):
+
+
+class EditPopup(object):
+  """Represents edit popup"""
```
Fixed
